guillotina/commands/server.py

```diff
@@ -1,20 +1,6 @@
-from aiohttp import web
+import uvicorn
```

Comment:
So we default with uvicorn but I guess we'll need docs/info on how to run guillotina with uvicorn/etc by the normal startup command for those `uvicorn guillotina:app_module_thingy` if possible.

Reply:
I think we have these options (options are written in order of preference (desc)):

1. Do somthing like `entrypoint = ServerCommand.run` in `guillotina/__init__.py` so any ASGI server can start guillotina with `G_CONFIG=./config.json uvicorn guillotina:entrypoint` (we can't pass custom arguments to our app when calling with uvicorn or any other server). With this approach the properties `host` and `port` won't be configurable from the `config.json`.
1. Hard-coding in the `ServerCommand` a list of ASGI servers, so we can launch the server with `g -c config.json serve --uvicorn` or `g -c config.json serve --hypercorn --host 0.0.0.0 --port 8080` (the `host` and `port` could be read from the config as well). But this would required do something like: 
```python
if arg == "--uvicorn":
    import uvicorn
    uvicorn.run(app, host=...)
elif arg == "--hypercorn":
    ...
```
3. Provide the functions `get_settings()` and `make_app(settings=)` and delegate to the user this logic. For example, by doing `app = make_app(settings=get_settings(file="config.json"))` in `app.py`, and then running the server with `uvicorn app:app` or by calling the server from code, like `if __name__ == "__main__": uvicorn.run(app, **params)` and running the file with `python app.py`

I'm not sure about the third option, could be tricky to implement it.

Thoughts?

Reply:
Can we do a combination of 1 and 2?

`#2` would include a default server to run with(say uvicorn) ootb but then you can also override it with maybe `--asgi-server=module.path.to.runner`.

`#2` would then depend on asgi providing a generic way to hook into for this--I'm not sure there isn't one ootb.

I think I would at the very least like an ootb implementation where the `serve` command worked with a default supported/tested implementation.

Thoughts?

Reply:
> #2 would include a default server to run with(say uvicorn) ootb but then you can also override it with maybe --asgi-server=module.path.to.runner.

The problem is there isn't a standard way to call an asgi server from code. 

* `uvicorn`:
```python
import uvicorn
uvicorn.run(app, host='0.0.0.0', port='0.0.0.0')
```
* `hypercorn`:
```python
# https://pgjones.gitlab.io/hypercorn/api_usage.html
from hypercorn.asyncio import serve
from hypercorn.config import Config

config = Config()
config.bind = ["localhost:8080"]  # As an example configuration setting
asyncio.run(serve(app, config))
```
* `daphne`:
```python
# Doesn't provide documentation on how to run the server from code
```

> I think I would at the very least like an ootb implementation where the serve command worked with a default supported/tested implementation.

I think we can provide an ootb implementation that works with `uvicorn` and `hypercorn`, and the user can select which server wants to run with `--asgi-server=hypercorn` (default to `uvicorn`). Something like:
```python
if arguments.asgi_server == "uvicorn":
    import uvicorn
    uvicorn.run(app, host=argument.host, port=argument.port)
elif arguments.asgi_server == "hypercorn":
    from hypercorn.asyncio import serve
    from hypercorn.config import Config
    config = Config()
    config.bind = [f"{argument.host}:{argument.port}"]
    asyncio.run(serve(app, config))
else:
    raise Exception(f"Server {arguments.asgi_server} not supported")
```

What do you think?

Reply:
Okay, I'm good with `--asgi-server`.

Maybe we can make sure to have good docs on creating app object to run with any server manually as well so users can integrate with another server before we have support for it. (or if they just want to have a different startup implementation)

Reply:
Done a first implementation. Suggestions are welcome :-)